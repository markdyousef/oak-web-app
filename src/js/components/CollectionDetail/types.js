type Seed = {
    id: string,
    content: string,
    cover: Object,
    creator: Object,
    labels: Array <Object>,
    likes: Array<number>,
    comments: Array<Object>
}

type Data = {
    loading: bool,
    refetch: Function,
    me?: {
        likedSeeds: ?Array<Object>
    },
    grove?: {
        id: string,
        name: string,
        description: ?string,
        cover: ?Object,
        stats: ?Object,
        labels: ?Array<Object>
    },
    seeds?: Array<Seed>
}
export type DefaultProps = {};
export type Props = {
    params: {
        collectionId: string
    },
    router: {
        push: Function,
        replace: Function
    },
    data: Data,
    remove: Function,
    removeCard: Function,
    likeCard: Function,
    unlikeCard: Function,
    setUpdate: (type: string, update: bool) => void,
    shouldUpdate: bool
};
export type State = {
    showEdit: bool,
    showDetail: bool,
    cards: Array<Seed>,
    sortKey: string,
    labels: Array<Object>,
    filterVals: Array<string>
};
