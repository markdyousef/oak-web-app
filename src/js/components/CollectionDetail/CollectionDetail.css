:root {
    --main-black-color: #000;
    --main-white-color: #fefeff;
    --main-orange-color: #ff9200;
    --main-red-color: #8b0000;
    --main-green-color: #005d6c;
}

.container {
    width: 100%;
}

.header {
    width: 100%;
    height: 300px;
    background-color: #fff;
    box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
    display: flex;
    flex-direction: column;
}

.back {
    height: 50px;
    width: 100px;
    display: flex;
    align-items: center;
    padding: 10px;
}

.grid {
    width: 100%;
    padding: 20px;
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
}

.card {
    transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);
    margin: 10px;
}

.card:hover {
    box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
}
