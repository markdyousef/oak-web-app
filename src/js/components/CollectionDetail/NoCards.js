// @flow
import React from 'react';
import styled from 'styled-components';
import Button from '../shared/Button';
import empty from '../../../img/comments-empty-state.svg';

const Empty = ({ onClick }:Object) => {
    console.log('no cards');
    return (
        <div>
            <img src={empty} alt="empty state" />
        </div>
    );
}

export default Empty;
