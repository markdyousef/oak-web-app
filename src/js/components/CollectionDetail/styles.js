import styled from 'styled-components';
import colors from '../../styles/colors';

export const Container = styled.div`
    width: 100%;
    height: 100%;
    overflow: auto;
`;

export const Header = styled.section`
    width: 100%;
    min-height: 260px;
    background-color: ${colors.white};
    padding: 48px 72px;
    border-bottom: 1px solid ${colors.lightGrey};
    display: flex;
    justify-content: space-between;
    flex-wrap: wrap;
`;

export const Info = styled.div`
    max-width: 500px;
    & h1 {
        font-size: 24px;
        font-weight: bold;
        padding-bottom: 24px;
        display: block;
        letter-spacing: -.04em;
    }
    & p {
        font-size: 18px;
        font-weight: normal;
        padding-bottom: 40px;
        display: block;
        line-height: 1.48;
    }
`;

export const Stats = styled.div`
    & div {
        display: inline-block;
        margin-right: 20px;
    }
    & h3 {
        font-size: 20px;
        font-weight: bold;
        padding-bottom: 8px;
        display: block;
    }
    & h5 {
        font-size: 14px;
        font-weight: 300;
        text-transform: uppercase;
        padding-bottom: 8px;
    }
`;

export const ButtonGroup = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    & button {
        margin-top: 5px;
    }
`;

export const Grid = styled.section`
    width: 100%;
    padding: 20px;
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
`;
