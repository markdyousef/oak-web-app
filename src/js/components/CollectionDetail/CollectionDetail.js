// @flow
import React, { Component } from 'react';
import Masonry from 'react-masonry-component';
import Card from '../Card';
import { SquareButton } from '../shared/Button';
import DotSpinner from '../shared/DotSpinner';
import CollectionDialog from '../../containers/CollectionDialogContainer';
import NoCards from './NoCards';
import CollectionToolbar from '../CollectionToolbar';

import {
    Container,
    Header,
    Info,
    Stats,
    ButtonGroup,
    Grid,
    masonStyles,
    Loading
} from './styles';

type Seed = {
    id: string,
    content: string,
    cover: Object,
    creator: Object,
    labels: Array <Object>,
    likes: Array<number>,
    comments: Array<Object>
}

type Data = {
    loading: bool,
    refetch: Function,
    me?: {
        likedSeeds: ?Array<Object>
    },
    grove?: {
        id: string,
        name: string,
        description: ?string,
        cover: ?Object,
        stats: ?Object,
        labels: ?Array<Object>
    },
    seeds?: Array<Seed>
}
type DefaultProps = {};
type Props = {
    params: {
        collectionId: string
    },
    router: {
        push: Function
    },
    data: Data,
    removeCard: Function,
    likeCard: Function,
    unlikeCard: Function
};
type State = {
    showEdit: bool,
    cards: Array<Seed>,
    sortKey: string,
    labels: Array<Object>,
    filterVals: Array<string>
};

class CollectionDetail extends Component<DefaultProps, Props, State> {
    static defaultProps: DefaultProps;
    props: Props;
    state: State;
    constructor() {
        super();
        this.state = {
            showEdit: false,
            cards: [],
            sortKey: 'date',
            labels: [],
            filterVals: []
        };
    }
    componentWillReceiveProps(nextProps:Props) {
        const { data: { loading, seeds, grove } } = nextProps;

        if (loading) return;

        if (seeds) {
            this.setState({ cards: seeds, sortKey: 'date' });
        }
        if (grove && grove.labels) {
            this.setState({ labels: grove.labels });
        }
    }
    onFilter = (key:string) => {
        const { filterVals, cards } = this.state;
        const { seeds } = this.props.data;
        let newFilters = filterVals;
        const active = filterVals.indexOf(key);
        if (active > -1) {
            newFilters = filterVals.filter(id => id !== key);
        } else {
            newFilters.push(key);
        }
        let filteredCards = seeds;
        if (newFilters.length > 0) {
            filteredCards = cards
                .filter((card) => {
                    if (card.labels) {
                        const hasLabel = card.labels.findIndex(label => label.id === key) > -1;
                        if (hasLabel) return card;
                    }
                    return null;
                })
                .filter(Boolean);
        }

        this.setState({
            filterVals: newFilters,
            cards: filteredCards
        });
    }
    onSortCards = (key: string, items?: Array<Object>) => {
        const { data: { seeds } } = this.props;
        const cards = (items) || this.state.cards;
        let sortedCards = [...cards];
        if (key === 'comments' || key === 'likes') {
            sortedCards.sort((a, b) => {
                if (a[key].length > b[key].length) {
                    return -1;
                }
                if (a[key].length < b[key].length) {
                    return 1;
                }
                return 0;
            });
        } else {
            sortedCards = seeds;
        }
        this.setState({ cards: sortedCards, sortKey: key });
    }
    removeCard = (cardId:string) => {
        const { removeCard, data } = this.props;
        removeCard(cardId)
            .then(() => data.refetch())
            .catch((err) => { throw err; });
    }
    handleLike = (cardId: string) => {
        const { likeCard, unlikeCard, data: { me, refetch } } = this.props;
        const isLiked = me && me.likedSeeds && me.likedSeeds.findIndex(item => item.id === cardId) > -1;
        if (isLiked) {
            unlikeCard(cardId)
                .then(() => refetch())
                .catch((err) => { throw err; });
        } else {
            likeCard(cardId)
                .then(() => refetch())
                .catch((err) => { throw err; });
        }
    }
    addCard = () => this.props.router.push(`/collection/${this.props.params.collectionId}/card`)
    showDialog = () => {
        const { showEdit } = this.state;
        const { data, router } = this.props;

        if (data.loading || !data.grove) return null;
        const { grove } = data;
        const cover = (grove.cover && grove.cover.urlThumb512) ? grove.cover : {};

        if (showEdit) {
            return (
                <CollectionDialog
                    close={() => {
                        data.refetch();
                        this.setState({ showEdit: false });
                    }}
                    name={data.grove.name}
                    description={data.grove.description}
                    editMode
                    id={data.grove.id}
                    router={router}
                    picture={cover.urlThumb512}
                    pictureId={cover.id}
                />
            );
        }
        return null;
    }
    renderCards = () => {
        const { data: { me, loading }, router, params } = this.props;
        const { cards } = this.state;

        if (loading || !me) return <div style={{ marginTop: '40px' }}><DotSpinner /></div>;

        if (cards.length > 0) {
            return (
                <Grid>
                    <Masonry
                        style={masonStyles}
                    >
                        {cards.map((item) => {
                            const isLiked = me.likedSeeds && me.likedSeeds.findIndex(card => card.id === item.id) > -1;
                            let content;
                            try {
                                content = JSON.parse(item.content)
                            } catch (e) {
                                content = null;
                            }
                            return (
                                <Card
                                    key={item.id}
                                    {...item}
                                    content={content}
                                    onShow={() => router.push(`/collection/${params.collectionId}/card/${item.id}`)}
                                    onRemove={() => this.removeCard(item.id)}
                                    onLike={() => this.handleLike(item.id)}
                                    showComments={() => router.push(`/collection/${params.collectionId}/card/${item.id}/comments`)}
                                    likes={item.likes.length}
                                    isLiked={isLiked}
                                    cover={item.cover}
                                />
                            );
                        }
                        )}
                    </Masonry>
                </Grid>
            );
        }
        return <NoCards onClick={this.addCard} />;
    }
    renderInfo = () => {
        const { data: { loading, grove } } = this.props;
        const info = Object.assign({}, grove);
        return (
            <Info>
                <h1>{(loading) ? <Loading /> : info.name}</h1>
                <h3>{(loading) ? <Loading /> : info.description}</h3>
                <Stats>
                    <div>
                        <h3>{(loading) ? <Loading /> : info.stats && info.stats.seeds}</h3>
                        <h5>{(loading) ? <Loading /> : 'Cards'}</h5>
                    </div>
                </Stats>
            </Info>
        );
    }
    render() {
        const { sortKey, labels, filterVals } = this.state;
        return (
            <Container>
                <Header>
                    {this.renderInfo()}
                    <ButtonGroup>
                        <SquareButton
                            onClick={this.addCard}
                            text="Add Card"
                            type="primary"
                        />
                        <SquareButton
                            onClick={() => this.setState({ showEdit: true })}
                            text="Edit Collection"
                        />
                        {this.showDialog()}
                    </ButtonGroup>
                </Header>
                <CollectionToolbar
                    onSort={this.onSortCards}
                    onFilter={this.onFilter}
                    active={sortKey}
                    labels={labels}
                    filters={filterVals}
                />
                {this.renderCards()}
            </Container>
        );
    }
}

export default CollectionDetail;
