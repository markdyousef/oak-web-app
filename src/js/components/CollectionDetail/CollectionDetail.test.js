import React from 'react';
import { shallow } from 'enzyme';
import CollectionDetail from './CollectionDetail';

it('renders correctly', () => {
    const shallowComponent = shallow(
        <CollectionDetail />
    );
    expect(shallowComponent).toMatchSnapshot();
});
