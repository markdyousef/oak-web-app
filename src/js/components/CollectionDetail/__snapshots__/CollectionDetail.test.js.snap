exports[`test renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <styled.div>
    <styled.section>
        <styled.div>
            <h1>
                <styled.div />
            </h1>
            <h3>
                <styled.div />
            </h3>
            <styled.div>
                <div>
                    <h3>
                        <styled.div />
                    </h3>
                    <h5>
                        <styled.div />
                    </h5>
                </div>
            </styled.div>
        </styled.div>
        <styled.div>
            <Unknown
                onClick={[Function]}
                text="Add Card"
                type="primary" />
            <Unknown
                onClick={[Function]}
                text="Edit Collection" />
        </styled.div>
    </styled.section>
    <Toolbar
        active="date"
        filters={Array []}
        labels={Array []}
        onFilter={[Function]}
        onSort={[Function]} />
    <div
        style={
            Object {
                "marginTop": "40px",
              }
        }>
        <Unknown />
    </div>
</styled.div>,
  "nodes": Array [
    <styled.div>
      <styled.section>
            <styled.div>
                  <h1>
                        <styled.div />
                  </h1>
                  <h3>
                        <styled.div />
                  </h3>
                  <styled.div>
                        <div>
                              <h3>
                                    <styled.div />
                              </h3>
                              <h5>
                                    <styled.div />
                              </h5>
                        </div>
                  </styled.div>
            </styled.div>
            <styled.div>
                  <Unknown
                        onClick={[Function]}
                        text="Add Card"
                        type="primary" />
                  <Unknown
                        onClick={[Function]}
                        text="Edit Collection" />
            </styled.div>
      </styled.section>
      <Toolbar
            active="date"
            filters={Array []}
            labels={Array []}
            onFilter={[Function]}
            onSort={[Function]} />
      <div
            style={
                  Object {
                        "marginTop": "40px",
                      }
            }>
            <Unknown />
      </div>
</styled.div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <CollectionDetail
        data={
                Object {
                        "loading": true,
                      }
        }
        likeCard={[Function]}
        params={Object {}}
        removeCard={[Function]}
        router={Object {}}
        unlikeCard={[Function]} />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": CollectionDetail {
        "_reactInternalInstance": [Circular],
        "addCard": [Function],
        "context": Object {},
        "handleLike": [Function],
        "onFilter": [Function],
        "onSortCards": [Function],
        "props": Object {
          "data": Object {
            "loading": true,
          },
          "likeCard": [Function],
          "params": Object {},
          "removeCard": [Function],
          "router": Object {},
          "unlikeCard": [Function],
        },
        "refs": Object {},
        "removeCard": [Function],
        "renderCards": [Function],
        "renderInfo": [Function],
        "showDialog": [Function],
        "state": Object {
          "cards": Array [],
          "filterVals": Array [],
          "labels": Array [],
          "showEdit": false,
          "sortKey": "date",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <styled.div>
          <styled.section>
                    <styled.div>
                              <h1>
                                        <styled.div />
                              </h1>
                              <h3>
                                        <styled.div />
                              </h3>
                              <styled.div>
                                        <div>
                                                  <h3>
                                                            <styled.div />
                                                  </h3>
                                                  <h5>
                                                            <styled.div />
                                                  </h5>
                                        </div>
                              </styled.div>
                    </styled.div>
                    <styled.div>
                              <Unknown
                                        onClick={[Function]}
                                        text="Add Card"
                                        type="primary" />
                              <Unknown
                                        onClick={[Function]}
                                        text="Edit Collection" />
                    </styled.div>
          </styled.section>
          <Toolbar
                    active="date"
                    filters={Array []}
                    labels={Array []}
                    onFilter={[Function]}
                    onSort={[Function]} />
          <div
                    style={
                              Object {
                                        "marginTop": "40px",
                                      }
                    }>
                    <Unknown />
          </div>
</styled.div>,
        "_debugID": 2,
        "_renderedOutput": <styled.div>
          <styled.section>
                    <styled.div>
                              <h1>
                                        <styled.div />
                              </h1>
                              <h3>
                                        <styled.div />
                              </h3>
                              <styled.div>
                                        <div>
                                                  <h3>
                                                            <styled.div />
                                                  </h3>
                                                  <h5>
                                                            <styled.div />
                                                  </h5>
                                        </div>
                              </styled.div>
                    </styled.div>
                    <styled.div>
                              <Unknown
                                        onClick={[Function]}
                                        text="Add Card"
                                        type="primary" />
                              <Unknown
                                        onClick={[Function]}
                                        text="Edit Collection" />
                    </styled.div>
          </styled.section>
          <Toolbar
                    active="date"
                    filters={Array []}
                    labels={Array []}
                    onFilter={[Function]}
                    onSort={[Function]} />
          <div
                    style={
                              Object {
                                        "marginTop": "40px",
                                      }
                    }>
                    <Unknown />
          </div>
</styled.div>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <CollectionDetail
    data={
        Object {
            "loading": true,
          }
    }
    likeCard={[Function]}
    params={Object {}}
    removeCard={[Function]}
    router={Object {}}
    unlikeCard={[Function]} />,
}
`;
